functionaltests/main_test.go

```diff
@@ -0,0 +1,226 @@
+	tf.Output("password", &escfg.Password)
+	tf.Output("kb_url", &escfg.KibanaURL)
+
+	t.Logf("created deployment %s", deploymentID)
+	fmt.Printf("%+v\n", escfg)
+
+	// Create APM Key
+	t.Log("creating API key")
+	ac, err := esclient.New(escfg)
+	require.NoError(t, err)
+
+	apikey, err := ac.CreateAPIKey(context.Background(),
+		t.Name(), -1, map[string]types.RoleDescriptor{},
+	)
+	require.NoError(t, err)
+
+	t.Log("ingest data")
+	// Ingest data through elastic/apm-perf docker image using apmtelemetrygen
+	// This is actually using an extracted telemetrygen as Go package.
+	// See https://github.com/elastic/apm-perf/pull/197
+	ingest(t, escfg.APMServerURL, apikey)
+
+	// Wait few seconds before proceeding to ensure ES indexed all our documents.
+	// In the tests I noticed that the datastream check could fail due to only
+	// 4 APM data streams being reported. Manual inspection revealed the
+	// correct number of data streams.
+	time.Sleep(20 * time.Second)
+
+	oldCount, err := ac.ApmDocCount(ctx)
+	require.NoError(t, err)
+
+	var dss []types.DataStream
+	// check data streams
+	t.Log("check data streams")
+	// GET _data_stream/*apm*
+	dss, err = ac.GetDataStream(ctx, "*apm*")
+	require.NoError(t, err)
+	checkDatastreams(t, checkDatastreamWant{
+		Quantity:         8,
+		PreferIlm:        false,
+		DSManagedBy:      "Data stream lifecycle",
+		IndicesPerDs:     1,
+		IndicesManagedBy: []string{"Data stream lifecycle"},
+	}, dss)
+
+	// Upgrade to 8.16.0
+	// FIXME: the update failed because it took more than 10m
+	t.Log("upgrade to 8.16.0")
+	require.NoError(t, tf.Apply(ctx,
+		name,
+		tfexec.Var(fmt.Sprintf("stack_version=%s", "8.16.0"))))
+
+	// check data
+	newCount, err := ac.ApmDocCount(ctx)
+	require.NoError(t, err)
+	assertDocCount(t, oldCount, newCount)
+
+	// check rollover happened
+	t.Log("check data streams after upgrade")
+	dss, err = ac.GetDataStream(ctx, "*apm*")
+	require.NoError(t, err)
+	// NO rollover here, expected? yes with lazy rollover
+	checkDatastreams(t, checkDatastreamWant{
+		Quantity:         8,
+		PreferIlm:        false,
+		DSManagedBy:      "Data stream lifecycle",
+		IndicesPerDs:     1,
+		IndicesManagedBy: []string{"Data stream lifecycle"},
+	}, dss)
+
+	// ingest more
+	t.Log("ingest more")
+	ingest(t, escfg.APMServerURL, apikey)
+
+	// check rollover
+	// Confirm datastreams are
+	// v managed by DSL if created after 8.15.0
+	// x managed by ILM if created before 8.15.0
+	t.Log("check data streams and verity lazy rollover happened")
+	dss, err = ac.GetDataStream(ctx, "*apm*")
+	require.NoError(t, err)
+	checkDatastreams(t, checkDatastreamWant{
+		Quantity:         8,
+		PreferIlm:        false,
+		DSManagedBy:      "Data stream lifecycle",
+		IndicesPerDs:     2,
+		IndicesManagedBy: []string{"Data stream lifecycle", "Data stream lifecycle"},
+	}, dss)
+
+	// check ES logs, there should be no errors
+	// TODO: how to get these from Elastic Cloud? Is it possible?
+
+	// cleanup
+	t.Log("cleanup")
```

Comment:
will this run always regardless of the failures? Just wondering if the `cleanup` step could be also set as another command we can use in the GH actions.

That's handy if something went wrong, so no leftovers are kept

Reply:
It was not, I've fixed it in [`4532fe5` (#14935)](https://github.com/elastic/apm-server/pull/14935/commits/4532fe5e5f9841fc46ee1067fb828ec204bb0991)

I added a flag to disable cleanup but it require manual action. In this way any automated test will automatically clean up but it can be overridden manually (for example for testing/troubleshooting).

Reply:
To follow up on running the cleanup manually, each test is expected to have a tfstate in it's own folder. So it's possible to "manually" (in CI)  run `terraform destroy` from that folder. 
In any case cleanup now always run in CI, so leftovers can only happen due to failures in deleting resources (I'm not sure how frequent this may be)